src/lib.rs

```diff
@@ -1036,6 +1140,58 @@ impl<'a> Event<'a> {
 
         (step, event)
     }
+
+    fn into_owned(&self) -> EventOwned {
+        EventOwned {
+            wd: self.wd.clone(),
+            mask: self.mask,
+            cookie: self.cookie,
+            name: self.name.map(OsStr::to_os_string),
+        }
+    }
+}
+
+
+/// An owned inotify event
+///
+/// This is the same as [`Inotify::Event`] except that it owns its `name`. This is used in the
+/// async implementation.
+///
+/// [`Inotify::Event`]: struct.Event.html
+#[derive(Clone, Debug)]
+pub struct EventOwned {
+    /// Identifies the watch this event originates from
+    ///
+    /// This [`WatchDescriptor`] is equal to the one that [`Inotify::add_watch`]
+    /// returned when interest for this event was registered. The
+    /// [`WatchDescriptor`] can be used to remove the watch using
+    /// [`Inotify::rm_watch`], thereby preventing future events of this type
+    /// from being created.
+    ///
+    /// [`WatchDescriptor`]: struct.WatchDescriptor.html
+    /// [`Inotify::add_watch`]: struct.Inotify.html#method.add_watch
+    /// [`Inotify::rm_watch`]: struct.Inotify.html#method.rm_watch
+    pub wd: WatchDescriptor,
+
+    /// Indicates what kind of event this is
+    pub mask: EventMask,
+
+    /// Connects related events to each other
+    ///
+    /// When a file is renamed, this results two events: [`MOVED_FROM`] and
+    /// [`MOVED_TO`]. The `cookie` field will be the same for both of them,
+    /// thereby making is possible to connect the event pair.
+    ///
+    /// [`MOVED_FROM`]: event_mask/constant.MOVED_FROM.html
+    /// [`MOVED_TO`]: event_mask/constant.MOVED_TO.html
+    pub cookie: u32,
+
+    /// The name of the file the event originates from
+    ///
+    /// This field is set only if the subject of the event is a file in a
+    /// watched directory. If the event concerns a file or directory that is
+    /// watched directly, `name` will be `None`.
+    pub name: Option<OsString>,
```

Comment:
Have you considered avoiding the duplication by making `Event` generic?

Something like this:
``` rust
pub struct Event<S> {
    // ...
    pub name: Option<S>,
}
```

Not sure if the lifetime of the `&'a OsStr` would cause trouble, but I guess it's worth a try.

Reply:
That does work. Commits squashed and updated to do this.